src/combinatorics/rado/auxiliary.lean

```diff
@@ -0,0 +1,144 @@
+def surj_on (f : α → β) (s : finset α) (t : finset β) : Prop := (s : set α).surj_on f t
+
+/-- The statement that a function `f` is bijective from a `finset` `s` to a `finset` `t` -/
+def bij_on (f : α → β) (s : finset α) (t : finset β) : Prop := (s : set α).bij_on f t
+
+end finset_versions
+
+/-!
+Auxiliary `finset` lemmas.
+-/
+
+section α
+
+variables {α : Type*} [decidable_eq α]
+
+lemma insert_sdiff_singleton_self {a : α} {s : finset α} (h : a ∈ s) : s = insert a (s \ {a}) :=
+begin
+  ext b,
+  simp only [mem_insert, mem_sdiff, mem_singleton],
+  by_cases H : b = a; simp [h, H],
+end
+
+lemma insert_sdiff_singleton_self' {a : α} {s : finset α} (h : a ∉ s) : s = insert a s \ {a} :=
+begin
+  rw insert_sdiff_of_mem _ (mem_singleton_self _),
+  exact (sdiff_singleton_not_mem_eq_self _ h).symm,
+end
+
+lemma card_insert_eq_card_iff (a : α) (s : finset α) : (insert a s).card = s.card ↔ a ∈ s :=
+begin
+  rw card_insert_eq_ite,
+  simp only [imp_false, ite_eq_left_iff, nat.succ_ne_self, not_not],
+end
+
+lemma card_insert_eq_card_add_one {a : α} {s : finset α} (h : a ∉ s) :
+  (insert a s).card = s.card + 1 :=
+by simp only [card_insert_eq_ite, h, if_false]
+
+end α
+
+section β
+
+variables {α β : Type*} [decidable_eq β]
+
+lemma image_eq_image_restrict_univ (s : finset α) (f : α → β) :
+  s.image f = univ.image (s.restrict f) :=
+by {ext, simp only [mem_image, univ_eq_attach, mem_attach, exists_true_left, subtype.coe_mk,
+                    restrict, finset.exists_coe]}
+
+lemma image_eq_range_coe (s : finset α) (f : α → β) :
+  (s.image f : set β) = set.range (f ∘ (coe : s → α)) :=
+begin
+  ext,
+  simp only [finset.coe_image, set.mem_image, finset.mem_coe, set.mem_range,
+             function.comp_app, finset.exists_coe, subtype.coe_mk, exists_prop],
+end
+
+end β
+
+section αβ
+
+variables {α β : Type*} [decidable_eq α] [decidable_eq β]
+
+lemma image_update {s : finset α} (f : α → β) {a : α} (h : a ∈ s) (b : β) :
+  s.image (function.update f a b) = (s \ {a}).image f ∪ {b} :=
+begin
+  ext c,
+  simp only [mem_image, exists_prop, mem_union, mem_sdiff, mem_singleton],
+  refine ⟨λ H, _, λ H, _⟩,
+  { obtain ⟨j, hj₁, hj₂⟩ := H,
+    rcases eq_or_ne j a with rfl | hji,
+    { exact or.inr (hj₂ ▸ function.update_same j b f), },
+    { exact or.inl ⟨j, ⟨hj₁, hji⟩, hj₂ ▸ (function.update_noteq hji b f).symm⟩, } },
+  { rcases H with ⟨j, ⟨hj₁, hj₂⟩, hj₃⟩ | rfl,
+    { exact ⟨j, hj₁, hj₃ ▸ function.update_noteq hj₂ b f⟩, },
+    { exact ⟨a, h, function.update_same a c f⟩, } }
+end
+
+lemma image_restrict_eq_image_image_coe (s : finset α) (f : α → β) (t : finset s) :
+  t.image (s.restrict f) = (t.image coe).image f :=
+by {ext, simp only [mem_image, exists_prop, subtype.coe_mk, exists_and_distrib_right,
+                    exists_eq_right, restrict, finset.exists_coe]}
+
+lemma bUnion_union (s t : finset α) (f : α → finset β) :
+  (s ∪ t).bUnion f = s.bUnion f ∪ t.bUnion f :=
```

Comment:
Please use `finset.sup` instead

Reply:
Can you explain why?

Reply:
When I'm dealing with (fin)sets, I find it much more natural to use intersection/union/subset than /inf/sup/le.

Reply:
There is [set.bUnion_union](https://leanprover-community.github.io/mathlib_docs/data/set/lattice.html#set.bUnion_union), btw.

Reply:
This lemma looks very reasonable to me; though I'd like to see the `bInter_inter` version too.

Reply:
`bInter` doesn't exist.